File: tests/test_solution.py
from __future__ import annotations

from lcpush.detect import detect
from lcpush.solution import (
    MAX_BYTES,
    brackets_balanced,
    byte_size,
    entry_point_names,
    line_count,
    matches_slug,
    normalize,
    reject_reason,
    soft_warnings,
)
from tests.conftest import CPP_SOLUTION, PY_SOLUTION


def test_normalize_converts_crlf_and_trims_trailing_space():
    assert normalize("a  \r\nb\t\r\n") == "a\nb\n"


def test_normalize_ensures_exactly_one_trailing_newline():
    assert normalize("x\n\n\n\n") == "x\n"
    assert normalize("x") == "x\n"


def test_normalize_preserves_content_byte_for_byte():
    assert normalize(CPP_SOLUTION) == CPP_SOLUTION


def test_normalize_empty():
    assert normalize("   \n\n") == ""


def test_line_and_byte_counts():
    assert line_count(PY_SOLUTION) == 8
    assert byte_size("héllo") == 6


def test_reject_empty():
    assert reject_reason("   \n ") == "Content is empty."


def test_reject_oversized():
    reason = reject_reason("x" * (MAX_BYTES + 1))
    assert reason is not None and "1MB" in reason


def test_accepts_normal_solution():
    assert reject_reason(PY_SOLUTION) is None


def test_entry_point_names_python():
    assert "twoSum" in entry_point_names(PY_SOLUTION)


def test_entry_point_names_cpp():
    assert "twoSum" in entry_point_names(CPP_SOLUTION)


def test_entry_point_names_javascript():
    js = "var lengthOfLongestSubstring = function(s) { return 0; };"
    assert entry_point_names(js) == ("lengthOfLongestSubstring",)


def test_brackets_balanced():
    assert brackets_balanced(PY_SOLUTION)
    assert brackets_balanced(CPP_SOLUTION)


def test_brackets_unbalanced_on_truncation():
    half = CPP_SOLUTION[: len(CPP_SOLUTION) // 2]
    assert not brackets_balanced(half)


def test_brackets_ignore_literals_and_comments():
    assert brackets_balanced('x = "{{{"  # )))\n')
    assert brackets_balanced("/* ( */ int a = 1;\n")


def test_matches_slug_camel_and_snake():
    assert matches_slug("twoSum", "two-sum")
    assert matches_slug("two_sum", "two-sum")
    assert matches_slug("TwoSum", "two-sum")
    assert not matches_slug("lengthOfLongestSubstring", "two-sum")


def test_matches_slug_is_permissive_without_a_slug():
    assert matches_slug("whatever", "")


def test_warning_on_unknown_language():
    warnings = soft_warnings("just some words here", detection=detect("just some words"))
    assert any("Could not identify" in w for w in warnings)


def test_warning_on_wrong_question():
    code = "var lengthOfLongestSubstring = function(s) { return 0; };\n"
    warnings = soft_warnings(code, slug="two-sum", title="Two Sum")
    assert any("lengthOfLongestSubstring" in w and "Two Sum" in w for w in warnings)


def test_no_wrong_question_warning_when_names_line_up():
    warnings = soft_warnings(PY_SOLUTION, slug="two-sum", title="Two Sum")
    assert not any("Wrong question" in w for w in warnings)


def test_warning_on_conflict_markers_and_todo():
    warnings = soft_warnings("<<<<<<< HEAD\nTODO: finish\n")
    assert any("conflict markers or TODOs" in w for w in warnings)


def test_warning_on_truncated_paste():
    half = CPP_SOLUTION[: len(CPP_SOLUTION) // 2]
    warnings = soft_warnings(half, detection=detect(half), slug="two-sum")
    assert any("Brackets are unbalanced" in w for w in warnings)


def test_clean_solution_has_no_warnings():
    assert soft_warnings(PY_SOLUTION, detection=detect(PY_SOLUTION), slug="two-sum") == ()

File: tests/test_solution.cpp
#include <catch2/catch_test_macros.hpp>

#include <algorithm>

#include "helpers/fixtures.hpp"
#include "lcpush/domain/detect.hpp"
#include "lcpush/domain/solution.hpp"

using namespace lcpush;
using lcpush::testing::kCppSolution;
using lcpush::testing::kPySolution;

namespace {

bool any_contains(const std::vector<std::string>& warnings, const std::string& needle) {
    return std::any_of(warnings.begin(), warnings.end(), [&](const std::string& w) {
        return w.find(needle) != std::string::npos;
    });
}

}  // namespace

TEST_CASE("normalize converts crlf and trims trailing space") {
    CHECK(solution::normalize("a  \r\nb\t\r\n") == "a\nb\n");
}

TEST_CASE("normalize ensures exactly one trailing newline") {
    CHECK(solution::normalize("x\n\n\n\n") == "x\n");
    CHECK(solution::normalize("x") == "x\n");
}

TEST_CASE("normalize preserves content byte for byte") {
    CHECK(solution::normalize(kCppSolution) == kCppSolution);
}

TEST_CASE("normalize empty") {
    CHECK(solution::normalize("   \n\n") == "");
}

TEST_CASE("line and byte counts") {
    CHECK(solution::line_count(kPySolution) == 8);
    CHECK(solution::byte_size("héllo") == 6);
}

TEST_CASE("reject empty") {
    auto reason = solution::reject_reason("   \n ");
    REQUIRE(reason.has_value());
    CHECK(*reason == "Content is empty.");
}

TEST_CASE("reject oversized") {
    std::string big(solution::kMaxBytes + 1, 'x');
    auto reason = solution::reject_reason(big);
    REQUIRE(reason.has_value());
    CHECK(reason->find("1MB") != std::string::npos);
}

TEST_CASE("accepts normal solution") {
    CHECK_FALSE(solution::reject_reason(kPySolution).has_value());
}

TEST_CASE("entry point names python") {
    auto names = solution::entry_point_names(kPySolution);
    CHECK(std::find(names.begin(), names.end(), "twoSum") != names.end());
}

TEST_CASE("entry point names cpp") {
    auto names = solution::entry_point_names(kCppSolution);
    CHECK(std::find(names.begin(), names.end(), "twoSum") != names.end());
}

TEST_CASE("entry point names javascript") {
    auto names = solution::entry_point_names(
        "var lengthOfLongestSubstring = function(s) { return 0; };");
    REQUIRE(names.size() == 1);
    CHECK(names[0] == "lengthOfLongestSubstring");
}

TEST_CASE("brackets balanced") {
    CHECK(solution::brackets_balanced(kPySolution));
    CHECK(solution::brackets_balanced(kCppSolution));
}

TEST_CASE("brackets unbalanced on truncation") {
    std::string full = kCppSolution;
    std::string half = full.substr(0, full.size() / 2);
    CHECK_FALSE(solution::brackets_balanced(half));
}

TEST_CASE("brackets ignore literals and comments") {
    CHECK(solution::brackets_balanced("x = \"{{{\"  # )))\n"));
    CHECK(solution::brackets_balanced("/* ( */ int a = 1;\n"));
}

TEST_CASE("matches slug camel and snake") {
    CHECK(solution::matches_slug("twoSum", "two-sum"));
    CHECK(solution::matches_slug("two_sum", "two-sum"));
    CHECK(solution::matches_slug("TwoSum", "two-sum"));
    CHECK_FALSE(solution::matches_slug("lengthOfLongestSubstring", "two-sum"));
}

TEST_CASE("matches slug is permissive without a slug") {
    CHECK(solution::matches_slug("whatever", ""));
}

TEST_CASE("warning on unknown language") {
    auto detection = detect::detect("just some words");
    auto warnings = solution::soft_warnings("just some words here", &detection);
    CHECK(any_contains(warnings, "Could not identify"));
}

TEST_CASE("warning on wrong question") {
    std::string code = "var lengthOfLongestSubstring = function(s) { return 0; };\n";
    auto warnings = solution::soft_warnings(code, nullptr, "two-sum", "Two Sum");
    bool hit = std::any_of(warnings.begin(), warnings.end(), [](const std::string& w) {
        return w.find("lengthOfLongestSubstring") != std::string::npos &&
               w.find("Two Sum") != std::string::npos;
    });
    CHECK(hit);
}

TEST_CASE("no wrong question warning when names line up") {
    auto warnings = solution::soft_warnings(kPySolution, nullptr, "two-sum", "Two Sum");
    CHECK_FALSE(any_contains(warnings, "Wrong question"));
}

TEST_CASE("warning on conflict markers and todo") {
    auto warnings = solution::soft_warnings("<<<<<<< HEAD\nTODO: finish\n");
    CHECK(any_contains(warnings, "conflict markers or TODOs"));
}

TEST_CASE("warning on truncated paste") {
    std::string full = kCppSolution;
    std::string half = full.substr(0, full.size() / 2);
    auto detection = detect::detect(half);
    auto warnings = solution::soft_warnings(half, &detection, "two-sum");
    CHECK(any_contains(warnings, "Brackets are unbalanced"));
}

TEST_CASE("clean solution has no warnings") {
    auto detection = detect::detect(kPySolution);
    CHECK(solution::soft_warnings(kPySolution, &detection, "two-sum").empty());
}
